src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java

```diff
@@ -172,17 +189,13 @@ public SnowflakeSinkServiceV2(
    * @param topicPartition TopicPartition passed from Kafka
    */
   @Override
-  public void startTask(String tableName, TopicPartition topicPartition) {
+  public void startPartition(String tableName, TopicPartition topicPartition) {
     // the table should be present before opening a channel so let's do a table existence check here
     createTableIfNotExists(tableName);
 
     // Create channel for the given partition
     createStreamingChannelForTopicPartition(
-        tableName,
-        topicPartition,
-        this.conn != null
-            && this.conn.hasSchemaEvolutionPermission(
-                tableName, connectorConfig.get(SNOWFLAKE_ROLE)));
+        tableName, topicPartition, tableName2SchemaEvolutionPermission.get(tableName));
```

Comment:
what if `tableName2SchemaEvolutionPermission.get(tableName)` is null? have we handled that case?

Reply:
`createTableIfNotExists` always make sure `tableName` exist. [Ref](https://github.com/snowflakedb/snowflake-kafka-connector/blob/ad5a34544ef706c3e7d9ec2247f057c226312dc6/src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java#L559C2-L559C2)

Reply:
I dont see `null` handling. 
So here is what I think we should do:
1. if table is present in cache, do a get. 
2. If not, call permission function and populate it in cache and do a get again. 
3. pass the boolean down from above fetches `createStreamingChannelForTopicPartition`. It is then guaranteed to be non null since your cache is <String, Boolean> and your parameter in function is boolean. 

Let me know if this makes sense.

Reply:
that check is fine. I am more concerned about the existence of tableName in hashmap. Not in db schema in snowflake. 

there is possibility that a new table/partition gets assigned to a task after rebalance. even though table must have created, your hashmap might not have that tableName.

Reply:
Talked offline, Alec is populating the cache in `createTableIfNotExists`. 
I will wait for the test to be added as mentioned earlier but otherwise looks very promising. Thanks